### PR TITLE
azure: Add mode dependency logic to deletion

### PR DESCRIPTION
**Pull request by hakman**
/cc @justinsb @rifelpet

### CODE REVIEW

File: pkg/resources/azure/azure.go [hunk 1]
```diff
@@ -161,16 +164,35 @@ func (g *resourceGetter) listVirtualNetworksAndSubnets(ctx context.Context) ([]*
 	return rs, nil
 }
 
-func (g *resourceGetter) toVirtualNetworkResource(vnet *network.VirtualNetwork) *resources.Resource {
+func (g *resourceGetter) toVirtualNetworkResource(vnet *network.VirtualNetwork) (*resources.Resource, error) {
+	var blocks []string
+	blocks = append(blocks, toKey(typeResourceGroup, g.resourceGroupName()))
+
+	nsgs := map[string]struct{}{}
```
Comment (johngmyers) [inline pkg/resources/azure/azure.go:171]: Why not use a `k8s.io/utils/set/set.Set[string]`?

Comment (hakman) [reply pkg/resources/azure/azure.go:171]: I think it's just an awkward way to not use memory. Should probably change to bool, to keep things simple?

Comment (hakman) [reply pkg/resources/azure/azure.go]: Interesting approach. Thanks.

File: pkg/resources/azure/azure.go [hunk 2]
```diff
@@ -319,16 +342,26 @@ func (g *resourceGetter) toVMScaleSetResource(vmss *compute.VirtualMachineScaleS
...
 			}
-			vnets[subnetID.VirtualNetworkName] = struct{}{}
-			subnets[subnetID.SubnetName] = struct{}{}
+			vnets.Insert(subnetID.VirtualNetworkName)
+			subnets.Insert(subnetID.SubnetName)
+			if ip.LoadBalancerBackendAddressPools != nil {
+				for _, lb := range *ip.LoadBalancerBackendAddressPools {
+					lbID, err := azure.ParseLoadBalancerID(*lb.ID)
+					if err != nil {
+						return nil, fmt.Errorf("parsing load balancer ID: %s", err)
```
Comment (justinsb) [inline pkg/resources/azure/azure.go:360]: Nit: should probably be `...: %w` (?)

File: upup/pkg/fi/cloudup/azure/azure_utils.go [hunk 3]
```diff
@@ -40,3 +40,119 @@ func ZoneToAvailabilityZoneNumber(zone string) (string, error) {
...
+		s.ResourceGroupName,
+		s.VirtualNetworkName,
+		s.SubnetName)
+}
+
+// ParseSubnetID parses a given subnet ID string and returns a SubnetID.
+func ParseSubnetID(s string) (*SubnetID, error) {
+	l := strings.Split(s, "/")
+	if len(l) != 11 {
+		return nil, fmt.Errorf("malformed format of subnet ID: %s, %d", s, len(l))
```
Comment (justinsb) [inline upup/pkg/fi/cloudup/azure/azure_utils.go:65]: Nit: I normally prefer to print with %q to highlight the "wrong" value, and I'm not sure that printing the length really adds any information (though I know it is handy during development!)

**Comment by justinsb** [comment]
Thanks @hakman 

/approve
/lgtm

File: upup/pkg/fi/cloudup/azure/azure_utils.go (hunk 3 above)
Comment (hakman) [reply upup/pkg/fi/cloudup/azure/azure_utils.go:65]: It's just copy/paste from the original code, wanted to be easy diff.

File: pkg/resources/azure/azure.go (hunk 2 above)
Comment (hakman) [reply pkg/resources/azure/azure.go:360]: It's just copy/paste from the original code, wanted to be easy diff.

**Pull request merged**